lib/fluent/plugin/in_tail.rb

```diff
@@ -498,6 +498,19 @@ def close_watcher_handles
 
     # refresh_watchers calls @tails.keys so we don't use stop_watcher -> start_watcher sequence for safety.
     def update_watcher(tail_watcher, pe, new_inode)
+      # TODO we should use another callback for this.
+      # To supress impact to existing logics, limit the case to `@follow_inodes`.
+      # We may not need `@follow_inodes` condition.
+      if @follow_inodes && new_inode.nil?
+        # nil inode means the file disappeared, so we only need to stop it.
+        @tails.delete(tail_watcher.path)
+        # Detach should be done immediately to avoid log duplication.
+        # (Other `detach_watcher_after_rotate_wait` may have the same problem.
+        #  We need another mechanism not to add duplicated TailWathcer with detaching TailWatcher.)
+        detach_watcher(tail_watcher, pe.read_inode)
```

Comment:
I fixed the comment.
This is related to https://github.com/fluent/fluentd/pull/4237#issuecomment-1633358632.
If we use `detach_watcher_after_rotate_wait` here, we can't avoid the problem that this PR wants to fix.

Reply:
Now I understand the bug, but cannot understand why we cannot `detach_watcher_after_rotate_wait` here yet.

Reply:
Because this fix mainly focuses on rotation cases, not cases where the file is actually removed.
In rotation cases, the new file is added very soon,
If waiting `rotate_wait`, the problem of https://github.com/fluent/fluentd/pull/4237#issuecomment-1633358632 definitely happens and causes log duplication.

Reply:
> * A: initial start
> 
> ```
> test.log
> ```
> 
>     * B: move the file
> 
> 
> ```
> test.log1
> ```
> 
>     * C: add the new current file
> 
> 
> ```
> test.log
> test.log1
> ```
> 
> When StatWatcher (or TimeTrigger) notifies between B and C, then the TailWatcher becomes an orphan, and a handle leak happens.

I consider the case where the notifying happens between B and C.
After C, the new file `test.log` will be added very soon.

It's an issue of which is faster, `StatWatcher`(or `TimeTrigger`) or rotation process, in intervals of less than a second.

Reply:
> If waiting rotate_wait, the problem of https://github.com/fluent/fluentd/pull/4237#issuecomment-1633358632 definitely happens and causes log duplication.

Oh, sorry. Not `definitely`.
This is depending on the timing of `refresh_watcher`.

If `refresh_watcher` add the new TailWatcher during `rotate_wait`, then log duplication occurs.

Reply:
You can see this problem by changing this line to `detach_watcher_after_rotate_wait ` and running `$ bundle exec rake test TESTOPTS='-t"TailInputTest::Update watchers for rotation with follow_inodes"'`.

Reply:
> If waiting `rotate_wait`, the problem of https://github.com/fluent/fluentd/pull/4237#issuecomment-1633358632 definitely happens and causes log duplication.

Ah I see, thanks.

It's a tough issue. It should be `detach_watcher_after_rotate_wait` when we resolve duplicate TailWatcher issue to support both rotation & removing.
One of solution or work around is setting `rotate_wait` to `0`. When all rotated files are expected to be tracked by `follow_inode`, `rotate_wait` isn't mostly needed. But whether or not it actually tails all rotated files is depend on user's setting, so we can't assume that all rotated files are tracked.

Reply:
> It should be detach_watcher_after_rotate_wait when we resolve duplicate TailWatcher issue to support both rotation & removing.

Yes! I thinks so too!

Reply:
> One of solution or work around is setting rotate_wait to 0. When all rotated files are expected to be tracked by follow_inode, rotate_wait isn't mostly needed. But whether or not it actually tails all rotated files is depend on user's setting, so we can't assume that all rotated files are tracked.

Yes. I think so too.

Reply:
Changing this to `detach_watcher_after_rotate_wait ` is possible for now.
At least, it will solve the handle leak problem, and duplicate logs are less likely to occur.

Reply:
> If waiting `rotate_wait`, the problem of https://github.com/fluent/fluentd/pull/4237#issuecomment-1633358632 definitely happens and causes log duplication.

Could you add this URL to the comment?

> Changing this to `detach_watcher_after_rotate_wait` is possible for now.

Hmm, it's hard to judge but I think it's better to change for consistency of the current code.
Users can work around it by `rotate_wait 0`.

Reply:
I see! I will change those points!

Reply:
Done as https://github.com/fluent/fluentd/compare/f87ab7b96cf5618730a781bd1dd4dbdbea6c1ab2..781e31319649e0a229dedd7abca8bc8d20a76034

Reply:
I have updated the issue comment as well.